### PR TITLE
Implement judgelib's shuffle judge

### PR DESCRIPTION
**Pull request by 5K1PY**

### CODE REVIEW

File: docs/batch_judge.md [hunk 1]
```diff
@@ -15,11 +15,21 @@ writing a custom judge is necessary. Set `out_check` in the config accordingly.
 ## Tokens judge
 
 A fast and versatile equality checker. Ignores whitespace, but not newlines.
+(Ignores newlines only at the end of file.)
+
+Tokens are separated by (possibly multiple) whitespace characters.
+For output to be correct, tokens need to be same as in the correct output file.
```
Comment (SvizelPritula) [inline docs/batch_judge.md:21]: Add articles:

```suggestion
For the output to be correct, the tokens need to be same as in the correct output file.
```

File: docs/batch_judge.md [hunk 2]
```diff
@@ -15,11 +15,21 @@ writing a custom judge is necessary. Set `out_check` in the config accordingly.
...
 
 You can customize the tokens judge with `tokens_ignore_newlines` or `tokens_ignore_case`.
 For comparing floats, set `tokens_float_rel_error` and `tokens_float_abs_error`.
 Details can be found in [config-documentation](/config-v3-documentation).
 
+## Shuffle judge
+
+Similarly to tokens judge compares output with correct output token-by-token.
```
Comment (SvizelPritula) [inline docs/batch_judge.md:29]: I find this sentence a bit hard to parse. Also, it lacks articles.

```suggestion
Similarly to the tokens judge, the shuffle judge compares the output with the correct output token-by-token.
```

File: pisek/task_jobs/tools.py [hunk 3]
```diff
@@ -132,7 +133,25 @@ def _run(self):
...
+    """Compiles the generic token-based judge."""
+
+    def __init__(self, env: Env, **kwargs) -> None:
+        super().__init__(
+            env=env, judge_name="token judge", judge="judge-token", **kwargs
+        )
+
+
+class PrepareShuffleJudge(PrepareJudgeLibJudge):
+    """Compiles the generic token-based judge."""
```
Comment (SvizelPritula) [inline pisek/task_jobs/tools.py:149]: This is the same description that `PrepareTokenJudge` has.

File: pisek/tools/judgelib/judge-shuffle.cc [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+	for (uint i=0; i < num_lines; i++) {
+		line *l = &lines[i];
+		printf("%08x %5d: ", l->hash, l->orig_line);
+		for (uint j=0; j < l->len; j++) {
+			tok *t = &l->toks[j];
+			printf("<%08x|%s>", t->hash, t->token);
+		}
+		printf("\n");
+	}
+#endif
```
Comment (SvizelPritula) [inline pisek/tools/judgelib/judge-shuffle.cc:280]: Nit: Do we want to leave in debugging code?

**Review by SvizelPritula — CHANGES_REQUESTED** [review]

File: config-v2-documentation [hunk 5]
```diff
@@ -88,6 +90,19 @@ tokens_float_abs_error=1e-30
...
 
+shuffle_mode=not_set
+# Only for out_check=shuffle
+# Which permutations are allowed:
+#   lines       Permutation of lines
+#   words       Permutation of words within each line
+#   lines_words Both lines and words
+#   tokens      Permutation of all tokens, ignoring line boundaries
+
+shuffle_ignore_case=0
```
[the remark this answers is not in the document]
Comment (5K1PY) [reply config-v2-documentation:101]: If we would have `ignore_case` for all `out_check`s, then it would make sense. But in this case it makes the config more clear having the two keys separate.

File: pisek/tools/judgelib/judge-shuffle.cc (hunk 4 above)
Comment (5K1PY) [reply pisek/tools/judgelib/judge-shuffle.cc:280]: Not having different version then upstream is higher priority.

File: pisek/task_jobs/tools.py [hunk 6]
```diff
@@ -137,7 +137,7 @@ def _run(self):
 
 
 class PrepareTokenJudge(PrepareJudgeLibJudge):
-    """Compiles the generic token-based judge."""
+    """Compiles the judge-token from judgelib."""
```
Comment (SvizelPritula) [inline pisek/task_jobs/tools.py:140]: This should be either "Compiles judge-token", or "Compiles the judge-token judge/program".

**Review by SvizelPritula — APPROVED** [review]
LGTM
